src/daemon/distribution.c

```diff
@@ -0,0 +1,402 @@
+/**
+ * collectd - src/daemon/distribution.c
+ * Copyright (C) 2020       Google LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Barbara bkjg Kaczorowska <bkjg at google.com>
+ */
+
+#include "distribution.h"
+#include <math.h>
+#include <pthread.h>
+
+typedef struct {
+  double max_boundary;
+  uint64_t counter;
+} bucket_t;
+
+struct distribution_s {
+  bucket_t *buckets;
+  double sum_gauges;
+  size_t num_buckets;
+  pthread_mutex_t mutex;
+};
+
+double *distribution_get_buckets_boundaries(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  double *boundaries = calloc(d->num_buckets, sizeof(double));
+
+  if (boundaries == NULL) {
+    return NULL;
+  }
+
+  for (size_t i = 0; i < d->num_buckets; ++i) {
+    boundaries[i] = d->buckets[i].max_boundary;
+  }
+
+  return boundaries;
+}
+
+uint64_t *distribution_get_buckets_counters(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  uint64_t *counters = calloc(d->num_buckets, sizeof(uint64_t));
+
+  if (counters == NULL) {
+    return NULL;
+  }
+
+  pthread_mutex_lock(&d->mutex);
```

Comment:
I use the mutex to return the bucket's counters of the distribution_t itself and not use the clone, because copying is extremely costly and holding mutex and returning counters of the original distribution_t is just faster.  

Mutux have to be hold by the clone function, so if I would clone distribution_t structure, then I could not have the mutex in the counters getter function.

Reply:
Thank you for the explanation!